accounting/accounting.go

```diff
@@ -0,0 +1,188 @@
+
+import (
+	"math"
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/LUSHDigital/core-lush/currency"
+)
+
+const (
+	// QuadruplePrecision describes 128 bits of precision for IEEE 754 decimals
+	// see: https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format
+	QuadruplePrecision uint = 128
+
+	// OctuplePrecision describes 256 bits of precision for for IEEE 754 decimals
+	// see: https://en.wikipedia.org/wiki/Octuple-precision_floating-point_format
+	OctuplePrecision uint = 256
+)
+
+var (
+	// Minimum rate of tax must be 0.
+	// As far as we know, there are as of writing, no markets with a negative sales tax.
+	min = itof(0)
+
+	// Baseline for creating a rate divisor.
+	base = itof(1)
+)
+
+// ValidateManyFloatsArePrecise tests that the given float64 arguments
+// have the desired precision, this is a convenience wrapper
+// around ValidateFloatIsPrecise.
+// NOTE: 2 digits past the dot is a business rule.
+func ValidateManyFloatsArePrecise(args ...float64) error {
+	for _, f := range args {
+		if err := ValidateFloatIsPrecise(f); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ValidateFloatIsPrecise ensures that a float64 value does not exceed
+// a precision of 2 digits past the decimal period. This ensures we do not
+// store incorrect currency data.
+// NOTE: 2 digits past the decimal period is a business rule.
+func ValidateFloatIsPrecise(f float64) error {
+	// parse the float, with the smallest number of digits necessary
+	parsed := strconv.FormatFloat(f, 'f', -1, 64)
+
+	// split the parsed number on the decimal period
+	parts := strings.Split(parsed, ".")
+
+	// in case of exact number...
+	if len(parts) == 1 {
+		return nil
+	}
+
+	// check the float's precision
+	if prec := len([]rune(parts[1])); prec > 2 {
+		return ErrFloatPrecision{
+			Value:     parsed,
+			Precision: prec,
+		}
+	}
+	return nil
+}
+
+// ToMinorUnit returns the currency data in it's minor unit, int64 format
+func ToMinorUnit(c currency.Currency, value float64) int64 {
+	return int64(math.Round(value * float64(c.Factor())))
+}
+
+// FromMinorUnit returns the currency data as a floating point from it's
+// minor currency unit format.
+func FromMinorUnit(c currency.Currency, value int64) float64 {
+	// fast path for currencies like JPY with a factor of 1.
+	if c.Factor() == 1 {
+		return float64(value)
+	}
+	var (
+		v = itof(value)
+		f = ftof(c.FactorAsFloat64())
+	)
+	f64, _ := newf().Quo(v, f).Float64()
+	return f64
+}
+
+// Exchange - Apply currency exchange rates to an amount.
+//
+// value - should always be given in the minor currency unit.
+// exchange - should always be given from the approved finance list.
+//
+// Rounding to the nearest even is a defined business rule.
+// Tills may round up to the nearest penny, but for reporting, the rule is
+// always to use banker's rounding.
+//
+// If unclear, see: // http://wiki.c2.com/?BankersRounding.
+func Exchange(c currency.Currency, value, exchange float64) float64 {
```

Comment:
I would really like us to only use integers or rats when it comes to the interface for dealing with currency values. This will make a lot more sense for us when using this for helping automate pricing data. I would even go so far as to make a new type that is `Amount` that holds its currency and its lowest common denominator integer value.

With this we could change this API to using those primitives.

```go
type Amount {
  Currency currency.Currency
  Amount int64
}

func MakeAmount(currency currency.Currency, value int64) Amount {}
func Exchange(amount Amount, currency currency.Currency, rate float64) Amount {}
```

Reply:
Done together over google meet